elfo-network/src/frame/lz4.rs

```diff
@@ -0,0 +1,138 @@
+//! This module implements custom LZ4 framing.
+//!
+//! Single LZ4 frame contains one or more envelopes compressed as a single
+//! LZ4 block along with some meta information like frame size. We do NOT
+//! use standard LZ4 framing.
+//!
+//! Structure of a single frame:
+//!           name           bits
+//! +---------------------------+----+
+//! | size of whole frame       | 32 |
+//! +---------------------------+----+
+//! | size of decompressed data | 32 |
+//! +---------------------------+----+
+//! | LZ4 block                 |rest|
+//! +---------------------------+----+
+//!
+//! All fields are encoded using LE ordering.
+
+// TODO: checksums.
+
+use byteorder::{LittleEndian, ReadBytesExt, WriteBytesExt};
+use std::io::Cursor;
+
+use eyre::{eyre, Result};
+
+use crate::codec::decode::DecodeState;
+
+pub(crate) struct LZ4Buffer {
+    buffer: Vec<u8>,
+}
+
+#[derive(Default)]
+pub(crate) struct DecompressStats {
+    /// How many compressed bytes were processed so far.
+    pub(crate) total_compressed_bytes: u64,
+    /// How many uncompressed bytes were produced during decompression so far.
+    pub(crate) total_uncompressed_bytes: u64,
+}
+
+#[derive(Default)]
+pub(crate) struct CompressStats {
+    /// How many uncompressed bytes were processed so far.
+    pub(crate) total_uncompressed_bytes: u64,
+    /// How many uncompressed bytes were produced during compression so far.
+    pub(crate) total_compressed_bytes: u64,
+}
+
+impl LZ4Buffer {
+    pub(crate) fn with_capacity(capacity: usize) -> Self {
+        Self {
+            buffer: Vec::with_capacity(capacity),
+        }
+    }
+
+    pub(crate) fn decompress_frame(
+        &mut self,
+        raw: &[u8],
+        stats: &mut DecompressStats,
+    ) -> Result<DecodeState<&[u8]>> {
+        if raw.len() < 4 {
+            return Ok(DecodeState::NeedMoreData {
+                total_length_estimate: 4,
+            });
+        }
+
+        let mut input = Cursor::new(raw);
+        let frame_size = input.read_u32::<LittleEndian>()? as usize;
+        if frame_size >= 200_000_000 {
+            return Err(eyre!("frame size is too big"));
+        }
+
+        if raw.len() < frame_size {
+            return Ok(DecodeState::NeedMoreData {
+                total_length_estimate: frame_size,
+            });
+        }
+
+        let decompressed_size = input.read_u32::<LittleEndian>()? as usize;
+        if decompressed_size >= 200_000_000 {
+            return Err(eyre!("decompressed size is too big"));
+        }
+
+        self.buffer.clear();
```

Comment:
We're losing 1/3 of LZ4's performance here. I think we should simply remove this line.

Ideally, the next line should be prepended with `if self.buffer.len() < decompressed_size` to avoid extra (nonsimdable) filling.